035-cruise-control-scaling-brokers.md

```diff
@@ -0,0 +1,113 @@
+
+## Current situation
+
+Currently, when Cruise Control is deployed alongside the Apache Kafka cluster, no automatic rebalancing happens when the cluster itself is scaled up or down.
+
+It means that, if the cluster is scaled up, no replicas will be moved to the newly added brokers.
+These brokers will only get partition replicas of newly created topics.
+Furthermore, by using the `KafkaRebalance` custom resource to run a rebalancing operation after the scaling up, it will move replicas across all the already existing brokers and not only to the new ones.
+It actually runs a full-rebalance with the drawback of more data movement, CPU usage and so on.
+
+In the same way, if the cluster is scaled down, the replicas hosted on the target brokers to remove are not moved out before the scaling down happens.
+Right now the scaling down is not blocked but doing it is not part of this proposal.
+This could result in some partitions having less ISR (In-Sync Replicas) than before the scaling down, with an impact on availability of the cluster.
+The current Cruise Control integration has no support, via the `KafkaRebalance` custom resource, to allow moving these replicas out of the target brokers before the scaling down happens.
+
+## Motivation
+
+When the Apache Kafka cluster is scaled up, it should be possible to use the Cruise Control integration to run a rebalancing by moving some replicas to the newly added brokers without involving the already existing ones.
+This would provide a faster rebalancing, not involving the entire cluster.
+
+Not stopping the scaling down operation when replicas are hosted on the target brokers to remove is not part of this proposal but it should be possible to move replicas out of the target brokers before the scaling down happens.
+It would allow to remove "empty" brokers which are not hosting any replicas, avoiding less ISR problem.
+
+## Proposal
+
+### Cruise Control add and remove brokers endpoints
+
+Cruise Control provides two specific HTTP REST endpoints to rebalance a cluster when scaling up or down:
+
+* [`/add_broker`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#add-a-list-of-new-brokers-to-kafka-cluster): this endpoint allows to trigger a rebalancing operation by moving replicas from the already existing brokers to the new ones but not moving them among existing brokers. It can be used right after the scaling up, for getting a proposal (as dryrun) and running the actual rebalancing.
+* [`/remove_broker`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#decommission-a-list-of-brokers-from-the-kafka-cluster): this endpoint allows to trigger a rebalancing operation by moving replicas out from the target brokers to remove, in order to leave them "empty" and not hosting any replicas. It can be used just before the scaling down, for getting a proposal (as dryrun) and running the actual rebalancing.
```

Comment:
Do these endpoints handle multiple brokers added / removed? Or do they handle only one as suggested by the endpoint name?

Reply:
Yeah the endpoint name is "singular", even the `brokerid` query parameter they handle is singular but in the end it gets a list of brokers.